internal/cache/file/downloader/job_manager_parallel_downloads_test.go

```diff
@@ -0,0 +1,199 @@
+// limitations under the License.
+
+package downloader
+
+import (
+	"context"
+	"crypto/rand"
+	"math"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/cache/util"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/config"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/storageutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func getMinObject(t *testing.T, objectName string, bucket gcs.Bucket) gcs.MinObject {
+	t.Helper()
+	ctx := context.Background()
+	minObject, _, err := bucket.StatObject(ctx, &gcs.StatObjectRequest{Name: objectName,
+		ForceFetchFromGcs: true})
+	if err != nil {
+		t.Fatalf("Error occured while stating the object: %v", err)
+	}
+	if minObject != nil {
+		return *minObject
+	}
+	return gcs.MinObject{}
+}
+
+func createObjectInStore(t *testing.T, objPath string, objSize int64, bucket gcs.Bucket) {
+	t.Helper()
+	objectContent := make([]byte, objSize)
+	_, err := rand.Read(objectContent)
+	if err != nil {
+		t.Fatalf("Error while generating random object content: %v", err)
+	}
+	_, err = storageutil.CreateObject(context.Background(), bucket, objPath, objectContent)
+	if err != nil {
+		t.Fatalf("Error while creating object in fakestorage: %v", err)
+	}
+}
+
+func configureFakeStorage(t *testing.T) storage.StorageHandle {
+	t.Helper()
+	fakeStorage := storage.NewFakeStorage()
+	t.Cleanup(func() { fakeStorage.ShutDown() })
+	return fakeStorage.CreateStorageHandle()
+}
+
+func configureCache(t *testing.T, maxSize int64) (*lru.Cache, string) {
+	t.Helper()
+	cache := lru.NewCache(uint64(maxSize))
+	cacheDir, err := os.MkdirTemp("", "gcsfuse_test")
+	if err != nil {
+		t.Fatalf("Error while creating the cache directory: %v", err)
+	}
+	t.Cleanup(func() { _ = os.RemoveAll(cacheDir) })
+	return cache, cacheDir
+}
+
+func createObjectInStoreAndInitCache(t *testing.T, cache *lru.Cache, bucket gcs.Bucket, objectName string, objectSize int64) gcs.MinObject {
+	t.Helper()
+	createObjectInStore(t, objectName, objectSize, bucket)
+	minObj := getMinObject(t, objectName, bucket)
+	fileInfoKey := data.FileInfoKey{
+		BucketName: storage.TestBucketName,
+		ObjectName: objectName,
+	}
+	fileInfo := data.FileInfo{
+		Key:              fileInfoKey,
+		ObjectGeneration: minObj.Generation,
+		FileSize:         minObj.Size,
+		Offset:           0,
+	}
+	fileInfoKeyName, err := fileInfoKey.Key()
+	if err != nil {
+		t.Fatalf("Error occurred while retrieving fileInfoKey: %v", err)
+	}
+	_, err = cache.Insert(fileInfoKeyName, fileInfo)
+	if err != nil {
+		t.Fatalf("Error occurred while inserting fileinfo into cache: %v", err)
+	}
+	return minObj
+}
+
+func TestParallelDownloads(t *testing.T) {
```

Comment:
would prefer to move these tests in downloader_test.go (just that the long name job_manager_parallel_downloads_test looks big)

Reply:
downloader_test uses some common setup and teardown which go against testing practices. I personally face readability challenges when I tried going through it. It's also counterintuitive that job_test.go is triggered by downloader_test.go. And downloader_test uses ogletest which we are moving away from.
So, I think a separate test is the better path here. I've renamed it to jm_parallel_downloads_test.go to make the name short.
In the longer run, we should fix the tests in downloader_test.go to be more readable and use less of common not-so-obvious setup and tear-down logic.

Reply:
Sure, we can work to make downloader_test.go and job_test.go separate and make it more readable. Fine with keeping how you have kept the tests now in this PR.